src/frontend/js/modules/table-of-content.js

```diff
@@ -0,0 +1,214 @@
+   *   <ul>
+   *     <li><a href="#"></a></li>
+   *     ...
+   *   </ul>
+   * </section>
+   */
+  createTableOfContent() {
+    this.tocElement = document.createElement('section');
+    this.tocElement.classList.add('table-of-content__list');
+
+    this.tags.forEach((tag) => {
+      const linkTarget = tag.querySelector('a').getAttribute('href');
+
+      const linkWrapper = document.createElement('li');
+      const linkBlock = document.createElement('a');
+
+      linkBlock.innerText = tag.innerText;
+      linkBlock.href = `${linkTarget}`;
+
+      linkWrapper.classList.add('table-of-content__list-item');
+
+      // additional indent for h3-h6
+      linkWrapper.classList.add(`table-of-content__list-item--${tag.tagName.toLowerCase()}`);
+
+      linkWrapper.appendChild(linkBlock);
+      this.tocElement.appendChild(linkWrapper);
+    });
+  }
+
+  /**
+   * Add table of content to the page
+   */
+  addTableOfContent() {
+    const header = document.createElement('header');
+    const container = document.createElement('section');
+
+    header.innerText = 'On this page';
+    header.classList.add('table-of-content__header');
+    container.appendChild(header);
+
+    container.classList.add('table-of-content');
+    container.appendChild(this.tocElement);
+
+    const tocWrapper = document.querySelector(this.tocWrapperSelector);
+
+    if (!tocWrapper) {
+      throw new Error('Table of content wrapper not found');
+    }
+
+    tocWrapper.appendChild(container);
+  }
+
+  /**
+   * Init intersection observer
+   */
+  initIntersectionObserver() {
+    const options = {
+      rootMargin: '-5% 0 -60%',
+    };
+
+    const callback = (entries) => {
+      entries.forEach((entry) => {
+        const target = entry.target;
+        const targetLink = target.querySelector('a').getAttribute('href');
+
+        /**
+         * Intersection state of block
+         *
+         * @type {boolean}
+         */
+        const isVisible = entry.isIntersecting;
+
+        /**
+         * Calculate scroll direction whith the following logic:
+         *
+         * DOWN: if block top is BELOW (coordinate value is greater) the intersection root top
+         * and block is NOT VISIBLE
+         *
+         * DOWN: if block top is ABOVE (coordinate value is less) the intersection root top
+         * and block is VISIBLE
+         *
+         * UP: if block top is ABOVE (coordinate value is less) the intersection root top
+         * and block is visible
+         *
+         * UP: if block top is BELOW (coordinate value is greater) the intersection root top
+         * and block is NOT VISIBLE
+         *
+         * Therefore we can use XOR operator for
+         * - is block's top is above root's top
+         * - is block visible
+         *
+         * @type {string}
+         */
+        const scrollDirection = ((entry.boundingClientRect.top < entry.rootBounds.top) ^ (entry.isIntersecting)) ? 'down' : 'up';
```

Comment:
why do we need to know a direction?

Seems like we just need to 
- clear previous active state from all items
- add an active state to the currently centered item.

Reply:
cause we had no section which covers all the content page. we just need to look at headers blocks only. then we need to know for each header when intersection occurs does in VISIBLE or not and does it places upper the target part of the screen.  

![image](https://user-images.githubusercontent.com/15259299/174800410-6c068eea-6cac-4a54-84d5-ce7e4907fdbc.png)

Reply:
and why do you need a scroll direction?

Reply:
> if `header_1` becomes **visible** and **scroll direction is DOWN** then we need to highlight `header_1` => CURRENT section

you have started reading a new section

--

> if `header_1` becomes **NOT visible** and **scroll direction is DOWN** then we need to highlight `header_1` => do nothing

you still watching the big section and header moved up

--

> if `header_1` becomes **NOT visible** and **scroll direction is UP** then we need to highlight `header_0` => PREVIOUS section

you scrolling up (back to the start of the page) and a header left the intersection field means you started reading the previous section

--

> if `header_1` becomes **visible** and **scroll direction is UP** then we need to highlight `header_1` => do nothing

you still scrolling up and find the header as a start of the section